### PR TITLE
[#586] Fix wrong dates due to offset of time

### PR DESCRIPTION
**Pull request by georgiana-b**
opentrials/opentrials#686

After a charming journey among the horrors of JS date handling, this is the best thing I could come up with.
Imho expressed in [this node-pg-types issue](https://github.com/brianc/node-pg-types/issues/50), the `Date`-only fields should **not** have any time information attached to them. However, seems that most JS developers (including the ones at `node-elasticsearch`) don't share my opinion so I had to settle with this for consistency.

### CODE REVIEW

File: config/index.js [hunk 1]
```diff
@@ -63,6 +63,28 @@ bookshelf.plugin('virtuals');
 bookshelf.plugin('pagination');
 config.bookshelf = bookshelf;
 
+// Overwrite node-pg-types date parser. See https://github.com/tgriesser/knex/issues/1750
+const pgTypes = require('pg').types;
+
+function parseDate(val) {
+  return new Date(Date.parse(val));
```
Comment (vitorbaptista) [inline config/index.js]: Is it possible for `parseDate()` to be called with null values as well? If so, it needs to handle this case, similarly to `parseDateArray`

File: config/index.js [hunk 2]
```diff
@@ -63,6 +63,28 @@ bookshelf.plugin('virtuals');
...
+    if (entry !== null) {
+      parsedEntry = parseDate(entry);
+    }
+    return parsedEntry;
+  });
+
+  return p.parse();
+}
+pgTypes.setTypeParser(1082, parseDate);
+pgTypes.setTypeParser(1182, parseDateArray);
```
Comment (vitorbaptista) [inline config/index.js]: When you move this to a separate file, please remove these magic numbers `1082` and `1182`, using a constant like `DATE_OID`, so it gets more readable.

File: test/config/pg_types.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  });
+
+  it('returns null when value is null', () => {
+    let testDate = null;
+    should.equal(pgTypes.getTypeParser(DATE_OID)(testDate), null);
+  });
+
+  it('raises when value is invalid', () => {
+    let testDate = 'not a date';
+    pgTypes.getTypeParser(DATE_OID).bind(null, testDate).should.throw(/^Invalid Date/);
```
Comment (georgiana-b) [inline test/config/pg_types.js]: I used `.bind()` here because if I just do `pgTypes.getTypeParser(DATE_OID)(testDate).should.throw(/^Invalid Date/)` the tests fail with this error:

```
  1) parseDate raises when value is invalid:
     Error: Invalid Date
      at parseDate (config/pg_types.js:9:650)
      at Context.it (test/config/pg_types.js:25:36)
```

**Comment by georgiana-b** [comment]
@nightsh Please take a look at this and approve if you think it's ok.

File: config/pg_types.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+function parseDateArray(value) {
+  if (!value) {
+    return null;
+  }
+
+  const p = pgTypes.arrayParser.create(value, (entry) => {
+    let parsedEntry = entry;
+    if (String(entry).toLowerCase() !== String(null)) {
```
Comment (nightsh) [inline config/pg_types.js]: To clarify why is this here: `null` value only enters the function as the string `'null'`. This condition works with either version of it.

File: test/config/pg_types.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    ]);
+  });
+
+  it('returns null when value is null', () => {
+    let testArray = null;
+    should(parseDateArray(testArray)).equal(null);
+  });
+
+  it('returns empty array when there are no values', () => {
+    let testArray = `{}`;
```
Comment (nightsh) [inline test/config/pg_types.js]: I would've named this `testPgArray`, just to avoid any confusions with "regular" arrays delimited by square brackets.

File: test/config/pg_types.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  });
+
+  it('returns null when value is null', () => {
+    let testDate = null;
+    should(parseDate(testDate)).equal(null);
+  });
+
+});
+
+describe('parseDateArray', () => {
```
Comment (nightsh) [inline test/config/pg_types.js]: `it('calls parseDate() on single array items')`

For when you pass it an array of a single item. So we'd avoid creating an interval (unless we really want to).

File: config/pg_types.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      parsedEntry = parseDate(entry);
+    } else {
+      parsedEntry = null;
+    }
+    return parsedEntry;
+  });
+
+  return p.parse();
```
Comment (nightsh) [inline config/pg_types.js]: A bit unclear what `p` means, unless you go back up to see it's a parser. Might be worth renaming it to `parser`.

File: test/config/pg_types.js (hunk 6 above)
Comment (georgiana-b) [inline test/config/pg_types.js]: An array of dates does not represent an interval in this case. It's just a normal array with date values.
